addons/account_edi/models/account_edi_document.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, api, _
+from psycopg2 import sql, OperationalError
+from contextlib import contextmanager
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class AccountEdiDocument(models.Model):
+    _name = 'account.edi.document'
+    _description = 'Electronic Document for an account.move'
+
+    name = fields.Char(related='attachment_id.name')
+    edi_format_id = fields.Many2one('account.edi.format')
+    edi_format_name = fields.Char(string='Format Name', related='edi_format_id.name')
+    move_id = fields.Many2one('account.move')
+    state = fields.Selection([('to_send', 'To Send'), ('sent', 'Sent'), ('to_cancel', 'To Cancel'), ('cancelled', 'Cancelled')])
+    error = fields.Html()
+    attachment_id = fields.Many2one('ir.attachment')
+
+    _sql_constraints = [
+        (
+            'unique_edi_document_by_move_by_format',
+            'UNIQUE(edi_format_id, move_id)',
+            'Only one edi document by move by format',
+        ),
+    ]
+
+    @contextmanager
+    def _with_lock(self):
+        """ Locks the documents in this recordset.
+            For example : lock the documents when sending them to web-services.
+        """
+        try:
+            with self.env.cr.savepoint():
+                self._cr.execute('SELECT * FROM account_edi_document WHERE id IN %s FOR UPDATE NOWAIT', [tuple(self.ids)])
+                yield
+        except OperationalError as e:
+            if e.pgcode == '55P03':
+                _logger.debug('Another transaction already locked documents rows. Cannot process documents.')
+            else:
+                raise e
+
+    def _create_jobs(self):
+        """Creates a list of jobs to be performed for these documents.
```

Comment:
`_create_edi_document_batches_to_process` ? to check

Reply:
I personally like the 'jobs', because it's not really batches. It can be, but sometimes there's only one move =>  I think it would be confusing to include 'batch' in the name because we have a method on edi.format that specify if it supports batching or not.

However, I renamed into _prepare_jobs because I agree that _create_jobs could infer that we are creating some kind of queue that will run immediately, in background or something.

I rewrote the docstring hoping that it would better explain what a job is.